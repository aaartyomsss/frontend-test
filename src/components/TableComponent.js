import React, { useState } from 'react';
import { Table, TableBody, TableCell, TableContainer, TableHead, TableRow, Button, CircularProgress, Typography, Paper } from '@material-ui/core';
import { makeStyles } from '@material-ui/core';
import { convertDate, sortByDate } from '../utils/helperFunctions';
import '../styles.css';
import ArrowUpwardIcon from '@material-ui/icons/ArrowUpward';
import ArrowDownwardIcon from '@material-ui/icons/ArrowDownward';

const createData = (obj) => {
    const { id, timestamp } = obj;
    const { oldValue, newValue } = obj.diff[0];
    return { id, timestamp, oldValue, newValue };
}

const useStyles = makeStyles((theme) => ({
    buttonAndLoading: {
        margin: '10px 0',
        [theme.breakpoints.down('sm')]: {
            height: '75%'
        }
    },
    error: {
        color: 'red',
        marginTop: 10,
        [theme.breakpoints.down('sm')]: {
            fontSize: 11
        }
    },
    tableHeader: {
        fontWeight: 'bold',
        [theme.breakpoints.down('sm')]: {
            fontSize: 12,
            margin: '10 0',
            padding: 0
        }
    },
    tableHeaderId: {
        fontWeight: 'bold',
        [theme.breakpoints.down('sm')]: {
            textAlign: 'right',
            fontSize: 12,
            margin: '10 0',
            padding: '0 15 0 0'
        }
    },
    tableRow: {
        [theme.breakpoints.down('sm')]: {
            fontSize: 12,
            margin: '5 0',
            padding: 0
        }
    },
    tableRowId: {
        [theme.breakpoints.down('sm')]: {
            textAlign: 'right',
            fontSize: 12,
            margin: '10 0',
            padding: '0 15 0 0'
        }
    },
    icon: {
        position: 'absolute',
    }
}))

const TableComponent = ({ dataArray, type, loading, fetch, errorMsg }) => {
    // Handling sorting by date 
    const [ filter, setFilter ] = useState('new');
    const classes = useStyles();

    // Formatting data
    const rows = sortByDate(dataArray.map(obj => createData(obj)), filter);

    return (
        <TableContainer component={Paper}>
            <Table>
                <TableHead>
                    <TableRow>
                        <TableCell 
                            id='switchFilter'
                            className={classes.tableHeader} 
                            onClick={() => setFilter(filter === 'new' ? 'old' : 'new')}>
                                <div>
                                    Date {filter === 'new' ? <ArrowUpwardIcon className={classes.icon}/> : <ArrowDownwardIcon className={classes.icon}/>}
                                </div>
                        </TableCell>
                        <TableCell className={classes.tableHeaderId}>{type === 'user' ? 'User ID' : 'Project ID'}</TableCell>
                        <TableCell className={classes.tableHeader}>Old Value</TableCell>
                        <TableCell className={classes.tableHeader}>New Value</TableCell>
                    </TableRow>
                </TableHead>
                <TableBody id='tableBody'>
                    {rows.map((row) => (
                        <TableRow key={row.id}>
                            <TableCell className={classes.tableRow}>{convertDate(row.timestamp)}</TableCell>
                            <TableCell className={classes.tableRowId}>{row.id}</TableCell>
                            <TableCell className={classes.tableRow}>{row.oldValue}</TableCell>
                            <TableCell className={classes.tableRow}>{row.newValue}</TableCell>
                        </TableRow>
                    ))}
                </TableBody>
            </Table>
            <div className='centerElementsDiv'>
                {errorMsg ? <Typography className={classes.error}>{errorMsg}</Typography> : null}
                <div>
                    {loading 
                        ? <CircularProgress className={classes.buttonAndLoading}/> 
                        : <Button 
                            id='fetch'
                            color='primary' 
                            variant='contained' 
                            onClick={() => fetch(type)} 
                            className={classes.buttonAndLoading}>{errorMsg ? 'Retry' : 'Load More'}</Button>}
                </div>
            </div>
        </TableContainer>
    )

}

export default TableComponent;