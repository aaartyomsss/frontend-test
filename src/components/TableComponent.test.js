import React from 'react';
import '@testing-library/jest-dom/extend-expect'
import { render, within } from '@testing-library/react'
import TableComponent from './TableComponent'
import api from '../lib/api/index'
import { mount } from 'enzyme'
import { usersDiff } from '../lib/api/data'

describe('Fetching mock data', () => {
    test('After initial fetching array has length 3', async () => {
        let data = []
        expect.assertions(1)
        const response = await api.getUsersDiff()
        expect(data.concat(response.data)).toHaveLength(3)
    })
})

describe('Table fetches and displays data', () => {
    let data = []
    let component
    beforeEach(() => {
        component = render(
            <TableComponent dataArray={data}/>
        )    
    })

    test('Renders table body', async () => {
        expect.assertions(1)
        const res = await api.getProjectsDiff()
        data.concat(res.data)
        expect(
            component.container.querySelector('#tableBody')
        ).toBeDefined()
        data = []
    })

    test('Table renders data properly', () => {
        data.forEach(obj => {
            const row = component.container.querySelector(`#${obj.id}`)
            const utils = within(row)
            expect(utils.getByText(obj.id)).toBeDefined()
        })
    })

})

describe('Table snapshot test', () => {
    const data = usersDiff
    const dataSortedNewestFirst = data.sort((a, b) => b.timestamp - a.timestamp)
    const dataSortedOldestFirst = data.sort((a, b) => a.timestamp - b.timestamp)
    const mockFunction = jest.fn()


    test('Filter by data is clickable', () => {
        const component = mount(<TableComponent dataArray={data} />)
        component.find('th#switchFilter').simulate('click')
        expect(component).toMatchSnapshot()
        component.unmount()
    })

    test('Fetch function is clickable', () => {
        const component = mount(<TableComponent dataArray={data} fetch={mockFunction}/>)
        component.find('button#fetch').simulate('click')
        expect(mockFunction).toHaveBeenCalled()
    })

    test('Initially table diplays newest first', () => {
        const component = mount(<TableComponent dataArray={data} />)
        expect(component.props().dataArray).toEqual(dataSortedNewestFirst)
    })

    test('Sorting by date works properly', () => {
        const component = mount(<TableComponent dataArray={data} />)
        component.find('th#switchFilter').simulate('click')
        expect(component.props().dataArray).toEqual(dataSortedOldestFirst)
    })

    test('After switching filter and fetching data, data is displayed according to the filter', () => {
        const initData = data.slice(0, 4)
        const addData = data.slice(4)
        const component = mount(<TableComponent dataArray={initData} />)
        component.find('th#switchFilter').simulate('click')
        component.setProps({ dataArray: initData.concat(addData)})
        expect(component.props().dataArray).toEqual(dataSortedOldestFirst)

    })
})