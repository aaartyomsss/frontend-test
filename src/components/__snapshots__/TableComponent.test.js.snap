// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table snapshot test Filter by data is clickable 1`] = `ReactWrapper {}`;
